### PR TITLE
Minor: Improve `datafusion-proto` documentation

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/8820

## Rationale for this change
While answering a question in the discord channel I realized the documentation for datafusion-proto could also be better (the documentation for datafusion-susbtrait really needs work, see https://github.com/apache/arrow-datafusion/issues/8820)

## What changes are included in this PR?
1. Make the usecase for protobuf clear
2. Improve documentation
3. Add links

## Are these changes tested?
Existing tests

## Are there any user-facing changes?
Just docs, no behavior changes

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Expr module contains core type definition for `Expr`.
+//! Logical Expressions: [`Expr`]
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:18]: This is a drive by cleanup

File: datafusion/proto/README.md [hunk 2]
```diff
@@ -17,11 +17,35 @@
   under the License.
 -->
 
-# DataFusion Proto
+# Apache Arrow DataFusion Proto
```
Comment (comphead) [inline datafusion/proto/README.md:20]: Not sure if it matters to also mentioned that proto backed by `prost`

Comment (alamb) [reply datafusion/proto/README.md:20]: Good idea, I added a reference

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Pull request merged**